pulp_python/app/tasks/upload.py

```diff
@@ -0,0 +1,80 @@
+import os
+from gettext import gettext as _
+import pkginfo
+import shutil
+import tempfile
+
+from pulpcore.plugin.models import Artifact, CreatedResource, Repository, RepositoryVersion
+from rest_framework import serializers
+
+from pulp_python.app.models import PythonPackageContent
+from pulp_python.app.utils import parse_project_metadata
+
+
+DIST_EXTENSIONS = {
+    ".whl": "bdist_wheel",
+    ".exe": "bdist_wininst",
+    ".egg": "bdist_egg",
+    ".tar.bz2": "sdist",
+    ".tar.gz": "sdist",
+    ".zip": "sdist",
+}
+
+DIST_TYPES = {
+    "bdist_wheel": pkginfo.Wheel,
+    "bdist_wininst": pkginfo.Distribution,
+    "bdist_egg": pkginfo.BDist,
+    "sdist": pkginfo.SDist,
+}
+
+
+def one_shot_upload(artifact_pk, filename, repository_pk=None):
+    """
+    One shot upload for pulp_python
+
+    Args:
+        artifact_pk: validated artifact
+        filename: file name
+        repository_pk: optional repository to add Content to
+    """
+    # iterate through extensions since splitext does not support things like .tar.gz
+    for ext, packagetype in DIST_EXTENSIONS.items():
+        if filename.endswith(ext):
+            # Copy file to a temp directory under the user provided filename, we do this
+            # because pkginfo validates that the filename has a valid extension before
+            # reading it
```

Comment:
Does it?  I know this came from the RPM plugin which did, I know nothing about pkginfo.

Reply:
Yes, pkginfo does, I verified it in the python shell.